### PR TITLE
Welcome to the future

### PR DESCRIPTION
**Pull request by bmelts**
tl;dr iOS 9-compatible; tests are XCTest now; GHUnit is totally gone, as are a few methods we weren't using.

The most interesting stuff is, thankfully, near the top - deprecation fixes, in particular the URL encoding-related ones. Apple doesn't provide a JavaScript-like character set for encoding the *whole* URL (since that's technically incorrect, each component of a URL has different restricted characters), so we should consider these methods deprecated (one is marked as such, the rest we still use and I'm not ready to flag them yet), but in the interim a manual character set of allowed characters is provided, following the relevant JavaScript methods' implementations.

The rest of this review is mostly taking a flamethrower to stuff. Mainly GHUnit, which has been completely purged; the tests are now XCTestCases, in a real test target.

Also: some stuff we weren't using that was causing trouble (or just felt crufty) has been deleted.

GHKit now builds with no warnings, and tests with no failures in like a fraction of the time! Huzzah!

### CODE REVIEW

File: Tests/NSDateUtilsTest.m [hunk 1]
```diff
@@ -8,54 +8,54 @@
...
-  GHAssertTrue([date gh_year] == 2012, nil);
+  XCTAssertTrue([date gh_day] == 1);
+  XCTAssertTrue([date gh_month] == 2);
+  XCTAssertTrue([date gh_year] == 2012);
 }
 
 - (void)testMonthSymbolsForFormat {
   NSArray *monthSymbols = [NSDate gh_monthSymbols];
-  GHTestLog([monthSymbols description]);
+  NSLog(@"%@", [monthSymbols description]);
```
Comment (SSheldon) [inline Tests/NSDateUtilsTest.m]: No need to call description, `%@` will do that for you, right?

File: Tests/NSInvocationUtilsTest.m [hunk 2]
```diff
@@ -92,113 +92,114 @@ - (void)testInvokeWithDelay {
...
-  GHTestLog(@"invokeTestingNestedCalled on main thread: %d", [NSThread isMainThread]);
-	GHAssertTrue([NSThread isMainThread], @"Should be on main thread; This is currently an expected failure!");
-	GHAssertTrue(n == 1, @"Should be equal to 1 but was %d", n);
+  NSLog(@"invokeTestingNestedCalled on main thread: %d", [NSThread isMainThread]);
+	XCTAssertTrue([NSThread isMainThread], @"Should be on main thread; This is currently an expected failure!");
+	XCTAssertTrue(n == 1, @"Should be equal to 1 but was %ld", (long)n);
 	invokeTestingNestedCalled_ = YES;
 }
-
+/* This doesn't currently test anything, because XCTests are run on the main thread
```
Comment (SSheldon) [inline Tests/NSInvocationUtilsTest.m]: Should we just nuke it then?

File: Classes/GHNSURL+Utils.h [hunk 3]
```diff
@@ -129,7 +131,7 @@
  @param s String to escape
  @result Encode string
  */
-+ (NSString *)gh_encode:(NSString *)s;
++ (NSString *)gh_encode:(NSString *)s DEPRECATED_MSG_ATTRIBUTE("Use -[NSString stringByAddingPercentEncodingWithAllowedCharacters:] with the appropriate NSCharacterSet for each part of the URL. You can't encode a whole URL properly.");
```
Comment (SSheldon) [inline Classes/GHNSURL+Utils.h]: It looks like we're barely using this, want to just delete it instead of deprecating it?

File: Classes/GHNSString+URL.h [hunk 4]
```diff
@@ -53,9 +57,11 @@
  
  @see NSURL#gh_encode:
  */
-- (NSString *)gh_URLEncode;
+- (NSString *)gh_URLEncode DEPRECATED_MSG_ATTRIBUTE("Use -[NSString stringByAddingPercentEncodingWithAllowedCharacters:] with the appropriate NSCharacterSet for each part of the URL. You can't encode a whole URL properly.");
```
Comment (SSheldon) [inline Classes/GHNSString+URL.h]: It looks like we're barely using this, want to just delete it instead of deprecating it?

File: Tests/NSInvocationUtilsTest.m (hunk 2 above)
Comment (bmelts) [inline Tests/NSInvocationUtilsTest.m]: Wasn't 100% comfortable doing so vs. just disabling it; on the other hand, who's going to be touching this code…

File: Classes/GHNSURL+Utils.h (hunk 3 above)
Comment (bmelts) [inline Classes/GHNSURL+Utils.h]: consider it done

File: Classes/GHNSString+URL.h (hunk 4 above)
Comment (bmelts) [inline Classes/GHNSString+URL.h]: consider it done

File: Tests/NSURLUtilsTest.m [hunk 5]
```diff
@@ -63,59 +46,18 @@ - (void)testQueryStringToDictionary {
...
-	XCTAssertEqualObjects(filtered, [NSURL URLWithString:@"https://user:pass@api.yelp.com:400/path?a=d&b=c#myfrag"]);
-}
-
 - (void)testQueryDictionaryWithArray {
 	NSArray *array1 = [NSArray arrayWithObjects:@"va", @"vb", @"vc", nil];
 	NSDictionary *dict = [NSDictionary dictionaryWithObjectsAndKeys:array1, @"key1", @"value2", @"key2", nil];
-	NSString *s = [NSURL gh_dictionaryToQueryString:dict sort:YES];
-	XCTAssertEqualObjects(s, @"key1=va%2Cvb%2Cvc&key2=value2");	
+	NSString *s = [NSURL gh_dictionaryToQueryString:dict];
+	XCTAssertEqualObjects(s, @"key1=va,vb,vc&key2=value2");
```
Comment (bmelts) [inline Tests/NSURLUtilsTest.m]: this behavior is the only thing i'm concerned about. while correct, this is technically a breaking change – it shouldn't impact anything like signatures (those were always calculated from the unescaped version anyway) but will, for example, mean potential cache misses: any URL with a collection query parameter like this would have been cached with a query string with escaped commas, and this will now fail to find that cached result.

the bigger concern imo is "does any API that consumes these collections _require_ them to be escaped? will anything break because of this?" my suspicion is no but i'm not 100% sure…

Comment (SSheldon) [inline Tests/NSURLUtilsTest.m]: Hmm good point, my concern would be around something like: someone using this method to encode the body of an `application/x-www-form-urlencoded` request. Previously, given input like `{"name": "foo+bar"}`, we'd produce a string like `"name=foo%2Bbar"`. After this change, it'll produce a string like `"name=foo+bar"`, which in `application/x-www-form-urlencoded` would actually be decoded as `{"name": "foo bar"}`.

File: Classes/GHNSURL+Utils.m [hunk 6]
```diff
@@ -62,86 +53,23 @@ + (NSArray *)gh_dictionaryToQueryArray:(NSDictionary *)queryDictionary sort:(BOO
     }
     
     if (!valueDescription) continue;
-    
-    if (encoded) key = [self gh_encodeComponent:key];
-    if (encoded) valueDescription = [self gh_encodeComponent:valueDescription];
-    [queryStrings addObject:[NSString stringWithFormat:@"%@=%@", key, valueDescription]];
+    NSURLQueryItem *item = [NSURLQueryItem queryItemWithName:key value:valueDescription];
```
Comment (SSheldon) [inline Classes/GHNSURL+Utils.m]: Like you've pointed out, there's a behavior change here ("foo+bar" no longer becomes "foo%2Bbar"). Do we need to preserver that old behavior? If not, I feel a little uncomfortable releasing this as the same method but without those old quirks, especially when its utility is questionable in a world with `NSURLComponents`.

If we don't need to preserve the old behavior, should we just delete `gh_dictionaryToQueryString:` and `gh_queryString`?

Comment (bmelts) [inline Classes/GHNSURL+Utils.m]: After much research, we do in fact need to preserve the old behavior – this is a breaking change that's incompatible with urllib, among others. Will migrate this back to the old behavior.

File: Tests/NSURLUtilsTest.m (hunk 5 above)
Comment (bmelts) [inline Tests/NSURLUtilsTest.m]: This ended up being a super valid concern! The previous behavior will be preserved.
